### ISSUE TITLE
Failed to join cluster creates stacktrace

### ISSUE THREAD

**Issue by electrical** [atomix/atomix#86]
```
Exception in thread "main" java.util.concurrent.CompletionException: java.lang.IllegalStateException: failed to join the cluster
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:769)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at io.atomix.copycat.server.state.ServerState.join(ServerState.java:598)
        at io.atomix.copycat.server.state.ServerState.lambda$join$59(ServerState.java:592)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at io.atomix.catalyst.transport.NettyConnection.lambda$handleResponseFailure$4(NettyConnection.java:172)
        at io.atomix.catalyst.util.concurrent.Runnables.lambda$logFailure$17(Runnables.java:20)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: failed to join the cluster
        ... 15 more
```

Perhaps this could be converted into a cleaner error message?
I also didn't see anything about it trying to retry to join the cluster?

**Comment by kuujo** [comment]
Well, really this is a function of the `CompletableFuture` API. The reason the stack trace is showing up is because code somewhere blocked on the future, and the future was completed exceptionally with that exception. But if you're referring to an example we can prettify it.

I do think some of the logging and log levels need to be cleaned up. It does log messages when retrying during joining the cluster, but they're debug messages right now. It should be logging a better message for retrying the join.

BTW where did this happen? I haven't documented this behavior yet (getting to it) but servers should only attempt to join the cluster if they're not already members of the cluster (listed in the `members` list) or already joined in the past. Servers now store their last known configuration to disk and use that if available. I guess my question is, did this happen in the leader election example?

**Comment by electrical** [comment]
I think this was with the standalone server example.
I'll do some more testing today and see if i can reproduce it with the latest master branch.

**Comment by kuujo** [comment]
@electrical make sure you clear out your log directories for all servers if you change configurations. I actually just experienced this issue and it was because I changed the server addresses but the logs for the old servers were still being used. Once a server has been started, Copycat will always load the configuration from disk on subsequent runs and will essentially ignore the user-provided configuration. This is necessary to allow clusters to evolve without user code having to handle that evolution, and it's also necessary to prevent partitions at startup from causing split brain.

**Comment by electrical** [comment]
@kuujo did some more testing and started the standalone example with clean dirs.

```
11:58:03.545 [copycat-server-devel01.dev.de01.ispavailability.com/192.168.100.20:5001] INFO  i.a.c.server.state.ServerContext - Server started successfully!
11:58:03.545 [copycat-server-devel01.dev.de01.ispavailability.com/192.168.100.20:5001] DEBUG i.a.copycat.server.state.ServerState - devel01.dev.de01.ispavailability.com/192.168.100.20:5001 - Attempting to join via localhost/127.0.0.1:5000
11:58:03.567 [copycat-server-devel01.dev.de01.ispavailability.com/192.168.100.20:5001] DEBUG i.a.copycat.server.state.ServerState - devel01.dev.de01.ispavailability.com/192.168.100.20:5001 - Failed to join localhost/127.0.0.1:5000
11:58:05.070 [copycat-server-devel01.dev.de01.ispavailability.com/192.168.100.20:5001] DEBUG i.a.copycat.server.state.ServerState - devel01.dev.de01.ispavailability.com/192.168.100.20:5001 - Attempting to join via localhost/127.0.0.1:5000
11:58:05.071 [copycat-server-devel01.dev.de01.ispavailability.com/192.168.100.20:5001] DEBUG i.a.copycat.server.state.ServerState - devel01.dev.de01.ispavailability.com/192.168.100.20:5001 - Failed to join localhost/127.0.0.1:5000
11:58:05.071 [copycat-server-devel01.dev.de01.ispavailability.com/192.168.100.20:5001] DEBUG i.a.copycat.server.state.ServerState - devel01.dev.de01.ispavailability.com/192.168.100.20:5001 - Cancelling join timeout
11:58:06.573 [copycat-server-devel01.dev.de01.ispavailability.com/192.168.100.20:5001] DEBUG i.a.copycat.server.state.ServerState - devel01.dev.de01.ispavailability.com/192.168.100.20:5001 - Attempting to join via localhost/127.0.0.1:5000
11:58:06.578 [copycat-server-devel01.dev.de01.ispavailability.com/192.168.100.20:5001] DEBUG i.a.copycat.server.state.ServerState - devel01.dev.de01.ispavailability.com/192.168.100.20:5001 - Failed to join localhost/127.0.0.1:5000
Exception in thread "main" java.util.concurrent.CompletionException: java.lang.IllegalStateException: failed to join the cluster
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:769)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at io.atomix.copycat.server.state.ServerState.join(ServerState.java:598)
        at io.atomix.copycat.server.state.ServerState.lambda$join$59(ServerState.java:592)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at io.atomix.catalyst.transport.NettyClient.lambda$null$15(NettyClient.java:88)
        at io.atomix.catalyst.util.concurrent.ThreadContext.lambda$execute$3(ThreadContext.java:125)
        at io.atomix.catalyst.util.concurrent.Runnables.lambda$logFailure$17(Runnables.java:20)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: failed to join the cluster
        ... 16 more
```

Even if i start a single node it gets to that error.
If i start the second one it gets in the same state and never forms a cluster.

**Comment by kuujo** [comment]
@electrical I'm sorry. I just realized the standalone server test hadn't been updated since fixing the split brain issue I mentioned above. Since that change, servers initially start based on whether their local `Address` is listed in the list of member `Address`es. If the local server's address is in the members list, it starts as a full member of the cluster. If the local server's address is not in the list, it attempts to join an existing cluster. This is necessary to determine whether to start a new cluster or join an existing cluster. The previous approach attempted to determine whether a cluster was already running and start a new cluster if not, but that had potential to cause split brain if multiple servers were joining and were partitioned from the cluster. Anyways...

The standalone server example does not include the local server address in the members list. This means every server is trying to join ever other server, but no server is running, so none of the joins succeed. I've pushed an updated version that uses the same code as the leader election example to configure the cluster.
